allennlp/data/iterators/multiprocess_iterator.py

```diff
@@ -0,0 +1,152 @@
+from typing import Iterable, Iterator, List, Optional
+import logging
+
+from torch.multiprocessing import Process, Queue, get_logger
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.util import lazy_groups_of
+from allennlp.data.instance import Instance
+from allennlp.data.iterators.data_iterator import DataIterator, TensorDict
+from allennlp.data.dataset import Batch
+from allennlp.data.vocabulary import Vocabulary
+
+logger = get_logger()  # pylint: disable=invalid-name
+logger.setLevel(logging.INFO)
+
+def _create_tensor_dicts(input_queue: Queue,
+                         output_queue: Queue,
+                         max_instances_in_memory: int,
+                         batch_size: int,
+                         vocab: Vocabulary) -> None:
+    """
+    Pulls at most ``max_instances_in_memory`` from the input_queue,
+    groups them into batches of size ``batch_size``, converts them
+    to ``TensorDict`` s, and puts them on the ``output_queue``.
+    """
+    instances: List[Instance] = []
+
+    def make_batches() -> None:
```

Comment:
Is there any reason we can't re-use an existing iterator here?  For many applications we'd want this to have more logic to do something like the bucket iterator, or optionally allow shuffling all of the instances read into memory before creating batches, etc.  Right now we will always read each file in the same order.